torch/distributed/fsdp/_wrap_utils.py

```diff
@@ -119,3 +127,138 @@ def _warn_on_overridden_mixed_precision(
         "These modules will be wrapped as separate FSDP instacnes with mixed "
         "precision disabled."
     )
+
+
+def _validate_frozen_params(
+    root_module: nn.Module,
+    modules_to_wrap: Set[nn.Module],
+    ignored_params: Set[nn.Parameter],
+    use_orig_params: bool,
+):
+    """
+    This checks that, given ``modules_to_wrap``, each module would manage
+    parameters that are uniformly frozen or non-frozen. This uniformity
+    requirement is strict for ``use_orig_params=False`` (hard error) and highly
+    recommended for ``use_orig_params=True`` (user warning).
+    """
+    topo_sorted_named_modules = _get_topo_sorted_named_modules(root_module)
+    reverse_topo_sorted_named_modules = reversed(topo_sorted_named_modules)
+    visited_modules: Set[nn.Module] = set()
+    for module_name, module in reverse_topo_sorted_named_modules:
+        if module in modules_to_wrap:
+            param_to_fqn = _get_managed_param_to_fqn(
+                module, ignored_params, visited_modules, module_name
+            )
+            frozen_param_fqns: List[str] = []
+            frozen_param_numel = 0
+            nonfrozen_param_fqns: List[str] = []
+            nonfrozen_param_numel = 0
+            for param, fqn in param_to_fqn.items():
+                if param.requires_grad:
+                    nonfrozen_param_fqns.append(fqn)
+                    nonfrozen_param_numel += param.numel()
+                else:
+                    frozen_param_fqns.append(fqn)
+                    frozen_param_numel += param.numel()
+            if len(frozen_param_fqns) > 0 and len(nonfrozen_param_fqns) > 0:
+                msg = f"{module_name} has both parameters with requires_grad=True and False."
+                if use_orig_params:
```

Comment:
It's non-obvious to me why `not use_orig_params` doesn't work. Without using orig params, we are reconstructing the params from the flat buffer every iteration. What is the `requires_grad` on this flat buffer? Clearly it has to be `requires_grad=True`, since we want to accumulate gradients into it. So how do I reconstruct a parameter that doesn't require grad from this requires grad parameter? The naive thing to do is detach (letting me get at the data without setting up a gradient edge.) Is there something in the reduce-scatter that chokes afterwards?

Reply:
For `use_orig_params=True`, mixing frozen and non-frozen does work functionally. It just may use memory higher than expected (hence a warning instead of an error).

For the frozen original parameters, we can just have them view into the flat parameter but still have `requires_grad=False`. FSDP takes care to not forward a gradient into the `.grad` attribute for those parameters.

Reply:
Yes, the `use_orig_params=True` case makes sense to me. I'm wondering about `use_orig_params=False` haha

Reply:
Ahh, I somehow missed the `not` in `not use_orig_params`.

I think the issue is for the optimizer. For `use_orig_params=False`, the optimizer sees the sharded `FlatParameter`s themselves. If the `FlatParameter` logically contains both frozen and non-frozen parameters, one thing we can do is enforce zeros in all gradient elements for frozen parameters part of the `FlatParameter`; however, this assumes that for the optimizer, a step with zero gradient is the same as no step (which I think is not always true). Nonetheless, maybe it is not unreasonable to just follow this behavior and set the `flat_param.requires_grad = any(orig_param.requires_grad for orig_param in flat_param._params)`.

Reply:
Ah, the interaction with optimizer certainly makes sense.

That being said, most optimizers, if you only ever have steps with zero gradients, it will be equivalent to not updating (e.g., momentum will always be zero). But you are losing most of the benefit from freezing in the first place here, so this is mostly an academic discussion.